pkg/istio-agent/xds-proxy.go

```diff
@@ -0,0 +1,329 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"golang.org/x/oauth2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/reflection"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultClientMaxReceiveMessageSize = 1024 * 1024 * 8
+	defaultInitialConnWindowSize       = 1024 * 1024 // default gRPC InitialWindowSize
+	defaultInitialWindowSize           = 1024 * 1024 // default gRPC ConnWindowSize
+)
+
+type XdsProxy struct {
+	stopChan             chan struct{}
+	downstreamListener   net.Listener
+	downstreamGrpcServer *grpc.Server
+	upstreamAddress      string
```

Comment:
Do we intend to use any other upstream other than Istiod - otherwise I would just call it "istiodAddress" to avoid any confusion? But if we want this XdsProxy to be generic and later we can move it out of this for any other purpose - calling upstream seems good

Reply:
Generic SGTM to me. We can comment its Istiod if we want